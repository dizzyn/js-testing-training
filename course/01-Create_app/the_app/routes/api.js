import cowsay from "cowsay";

export function uppercase(str) {
  return String(str).toUpperCase();
}

export function cow(str) {
  return cowsay.say({
    text: str,
  });
}

export default function api(req, res) {
  const action = req.query.action;
  const str = req.query.str;

  if (!str) {
    res.status(400);
    res.send(`str not given`);
    return;
  }

  if (!action) {
    res.status(400);
    res.send(`action not given`);
    return;
  }

  if (action === "uppercase") {
    res.send(uppercase(str));
  } else if (action === "cow") {
    res.send(cow(str));
  } else {
    res.status(400);
    res.send(`Unknown action '${action}'`);
    return;
  }
}
