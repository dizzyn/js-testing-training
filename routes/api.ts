import { Request, Response } from "express";
// @ts-ignore
import cowsay from "cowsay";
// @ts-ignore
import text2png from "text2png";

export function uppercase(str: string) {
  return String(str).toUpperCase();
}

export function asciiart(str: string) {
  return cowsay.say({
    text: str,
  });
}

export function image(str: string) {
  return String(str).toUpperCase();
}

export default function api(req: Request, res: Response) {
  const action = req.query.action;
  const format = req.query.format;
  const str = req.query.str as string;

  if (!str) {
    res.status(400);
    res.send(`str not given`);
    return;
  }

  if (!format) {
    res.status(400);
    res.send(`format not given`);
    return;
  }

  if (!action) {
    res.status(400);
    res.send(`action not given`);
    return;
  }

  let newStr;

  if (action === "uppercase") {
    newStr = uppercase(str);
  } else if (action === "asciiart") {
    newStr = asciiart(str);
  } else {
    res.status(400);
    res.send(`Not allowed action '${action}'`);
    return;
  }

  if (format === "string") {
    res.send(newStr);
  } else if (format === "image") {
    res.set("Content-Type", "image/png");
    res.send(text2png(newStr, { color: "white" }));
  } else {
    res.status(400);
    res.send(`Not allowed format '${action}'`);
    return;
  }
}
