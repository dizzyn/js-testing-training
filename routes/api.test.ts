import api, { uppercase, asciiart } from "./api";
import { Request, Response } from "express";

it("uppercase", () => {
  expect(uppercase("A")).toBe("A");
  expect(uppercase("a")).toBe("A");
});

it("asciiart", () => {
  expect(asciiart("A")).toMatchSnapshot();
});

const request = { query: {} } as Request;
const response = ({
  status: jest.fn(),
  send: jest.fn(),
} as any) as Response;

function testHandler(query: any, status?: number) {
  jest.resetAllMocks();
  request.query = query;
  api(request, response);
  if (status) {
    expect(response.status).toBeCalledWith(status);
  }

  // @ts-ignore
  return response.send.mock.calls[0][0] as any;
}

it("Missing parameters", () => {
  expect(testHandler({}, 400)).toBe("str not given");
  expect(testHandler({ str: "x" }, 400)).toBe("format not given");
  expect(testHandler({ str: "x", format: "string" }, 400)).toBe(
    "action not given"
  );
  expect(testHandler({ str: "x", action: "ascii" }, 400)).toBe(
    "format not given"
  );
});

it.only("Uppercase", () => {
  expect(testHandler({ str: "a", action: "uppercase", format: "string" })).toBe(
    "A"
  );
});

it.only("Ascii", () => {
  expect(
    testHandler({ str: "a", action: "asciiart", format: "string" })
  ).toBeDefined();
});
