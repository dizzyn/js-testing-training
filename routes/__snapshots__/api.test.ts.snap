// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asciiart 1`] = `
" ___
< A >
 ---
        \\\\   ^__^
         \\\\  (oo)\\\\_______
            (__)\\\\       )\\\\/\\\\
                ||----w |
                ||     ||"
`;
